source/server/admin/stats_handler.cc

```diff
@@ -220,29 +245,70 @@ StatsHandler::statsAsJson(const std::map<std::string, uint64_t>& all_stats,
     stats_array.push_back(ValueUtil::structValue(stat_obj));
   }
 
-  ProtobufWkt::Struct histograms_obj;
-  auto* histograms_obj_fields = histograms_obj.mutable_fields();
-
   ProtobufWkt::Struct histograms_obj_container;
   auto* histograms_obj_container_fields = histograms_obj_container.mutable_fields();
-  std::vector<ProtobufWkt::Value> computed_quantile_array;
 
-  bool found_used_histogram = false;
+  // Display bucket data if histogram_buckets query parameter is used, otherwise output contains
+  // quantile summary data.
+  if (!histogram_buckets_value.has_value()) {
+    statsAsJsonQuantileSummaryHelper(*histograms_obj_container_fields, all_histograms, used_only,
+                                     regex);
+  } else if (histogram_buckets_value.value() == "cumulative") {
+    statsAsJsonCumulativeHistogramBucketsHelper(*histograms_obj_container_fields, all_histograms,
+                                                used_only, regex);
+  } else {
+    // "disjoint" should be only possible value at this point. Other values should already have
+    // caused a bad request response.
+    ASSERT(histogram_buckets_value.value() == "disjoint");
```

Comment:
ditto turn histogram_buckets_value into an enum and use a switch with no default.

Reply:
Done